### PR TITLE
Gjoranv/wait for deconstruct upon shutdown

### PR DESCRIPTION
**Pull request by gjoranv**
This is to prevent uninstalling bundles and ClassNotFound exceptions when the container is shutting down.

FYI: @bjorncs

### CODE REVIEW

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java [hunk 1]
```diff
@@ -61,9 +81,24 @@ public void deconstruct(List<Object> components, Collection<Bundle> bundles) {
...
         }
-        if (! destructibleComponents.isEmpty() || ! bundles.isEmpty())
-            executor.schedule(new DestructComponentTask(destructibleComponents, bundles),
+        if (!destructibleComponents.isEmpty() || !bundles.isEmpty()) {
+            var task = executor.schedule(new DestructComponentTask(destructibleComponents, bundles),
                               delay.getSeconds(), TimeUnit.SECONDS);
+            if (mode.equals(Mode.SHUTDOWN)) {
+                // Wait for deconstruction to finish
+                try {
+                    log.info("Waiting up to " + SHUTDOWN_DECONSTRUCT_TIMEOUT.toSeconds() + " seconds for all components to deconstruct.");
```
Comment (jonmv) [inline container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Isn't there also a 50s timeout on deconstruct, somewhere further up?

Comment (gjoranv) [reply container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Yes, you're right. That seems a bit harsh. Perhaps we should increase it to at least 90 sec, and set this one to ~80?

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java [hunk 2]
```diff
@@ -35,13 +39,29 @@
...
 
+    private static final Duration SHUTDOWN_DECONSTRUCT_TIMEOUT = Duration.ofMinutes(10);
+
+    public enum Mode {
+        RECONFIG,  // Delay deconstruction to allow old components to finish processing in-flight requests.
+        SHUTDOWN   // The container is shutting down. Start deconstructing immediately, and wait until all components
+                   // are deconstructed, to prevent shutting down while deconstruct is in progress.
+    }
+
+    // TODO: make private again
```
Comment (jonmv) [inline container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Yes, was made visible for testing. All tests can use the synchronous shutdown instead :)

**Review by jonmv — APPROVED** [review]
Do check the shutdown timeout.

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java (hunk 1 above)
Comment (jonmv) [reply container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: I think 45s is good!

**Comment by jonmv** [comment]
I'm too slow, I see :) This is very good!

**Review by jonmv — APPROVED** [review]

**Pull request merged**